scrapy_poet/overrides.py

```diff
@@ -89,26 +91,27 @@ def from_crawler(cls, crawler: Crawler) -> Crawler:
         return cls(crawler.settings.getlist("SCRAPY_POET_OVERRIDES", []))
 
     def __init__(self, rules: Optional[Iterable[RuleFromUser]] = None) -> None:
-        self.rules: List[OverrideRule] = []
-        self.matcher: Dict[Callable, URLMatcher] = defaultdict(URLMatcher)
+        self.rules: List[ApplyRule] = []
+        self.matcher: Dict[PageObject, URLMatcher] = defaultdict(URLMatcher)
         for rule in rules or []:
             self.add_rule(rule)
-        logger.debug(f"List of parsed OverrideRules:\n{self.rules}")
+        logger.debug(f"List of parsed ApplyRules:\n{self.rules}")
 
     def add_rule(self, rule: RuleFromUser) -> None:
         if isinstance(rule, (tuple, list)):
             if len(rule) != 3:
                 raise ValueError(
-                    f"Invalid overrides rule: {rule}. Rules as tuples must have "
+                    f"Invalid rule: {rule}. Rules as tuples must have "
                     f"3 elements: (1) the pattern, (2) the PO class used as a "
                     f"replacement and (3) the PO class to be replaced."
```

Comment:
I'm proposing to remove (or at least set for deprecation) the current feature of declaring the rules via triplets:

```python
"SCRAPY_POET_OVERRIDES": [
    ("toscrape.com", BTSBookListPage, BookListPage),
    ("toscrape.com", BTSBookPage, BookPage),
    ("bookpage.com", BPBookListPage, BookListPage),
    ("bookpage.com", BPBookPage, BookPage)
]
```

This could be a bit confusing to users with the introduction of `to_return` since either `instead_of` and `to_return` can now be `None`.

Approaches:
1. We could continue supporting the triplets with its current behavior but set a deprecation warning each time. If the user wants to use `to_return`, they can use `ApplyRule` to declare it.
2. Fully remove the triplets interface and for the user to declare `ApplyRule`.

I'm a bit inclined with approach 2 since I'm not a fan of the triplet interface. It's not clear which of the 2nd or 3rd element is the one assigned to `use` or `instead_of`.

Reply:
I'm fine with doing (2), but before going this way it could make sense to make ApplyRule easier to create.
Currently the example would look like this, right?

```py
("toscrape.com", BTSBookListPage, BookListPage)  # was
ApplyRule(for_patterns=Patterns(["toscrape.com"]), use=BTSBookListPage, instead_of=BookListPage)  # now
```

Most of it is fine, but `for_patterns=Patterns(["toscrape.com"])` is quite a bit worse than `"toscrape.com"`.
I wonder if we should make a shortcut in ApplyRule, so that the following works:

```py
ApplyRule("toscrape.com", use=BTSBookListPage, instead_of=BookListPage)
```

Reply:
Good idea! Yes, an attrs converter would solve this nicely. :)

Reply:
Implemented in `web-poet`: https://github.com/scrapinghub/web-poet/pull/84/commits/33a48ac967176218dd2552c97e916a9d54279212. 🎉 

I'll mark this on the backlog for now as integrating the new `to_return` takes more priority.